### ISSUE TITLE
Make the autohttps pod automatically re-attempt TLS cert acquisition on failure during startup somehow

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#2150]
Our autohttps pod that does TLS termination (encrypt/decrypt outbound/inbound HTTPS traffic) is using running the `traefik` docker image with the Traefik software configured to our needs. Treafik in turn relies on LEGO or the Let's Encrypt Go-lang ACME client to request/renew a TLS certificate.

The enhancement that is requested with this issue is to make sure that there is no need for a user to manually restart the autohttps pod in order to acquire a TLS certificate, which i think is needed if something goes wrong during when it attempts to do so during startup.

It can reduce the complexity to get this chart up and running with TLS if we help this become robust and automated, and it can help us avoid issues in CI tests that currently have steps to really ensure our ACME server is up and running before installing our helm chart so that it can succeed on first and only attempt.

### Implementation ideas
I think this can be a good situation to use a `startupProbe` or `livenessProbe`, but the challenge is that we need to have a clear signal to indicate if we succeeded or not to acquire the TLS certs, so we would probably need to use an `exec` probe (executes a command in the container) rather than a `http` probe (sends a web request) to do so.

It would be great if we could get a direct feedback from traefik on its status in the TLS cert acquisition rather than trying probing a file in the file system or similar though. I'm not sure what can be done.

### Action points to resolve this
- Propose a implementation strategy on how to resolve this
- Describe why you think it makes sense and what your deliberations are regarding the strategy
- Develop and open a PR

### Contribute with research!
- Verify if traefik still doesn't support re-attempting failed ACME challenges on startup.
- Explore if a command as executed form a terminal in the container traefik is running can detect if the initial ACME challenge has failed or not - it can be used as a k8s startupProbe to help us automatically restart the container on failure to re-attempt the ACME challenge.

**Comment by pvanliefland** [comment]
I might be wrong here, but while debugging the issue (for a new deployment - from scratch- it seemed that the autohttps pod was trying to acquire the certificate "too soon" (the load balancer was not ready at the time?)

**Comment by consideRatio** [comment]
> I might be wrong here, but while debugging the issue (for a new deployment - from scratch- it seemed that the autohttps pod was trying to acquire the certificate "too soon" (the load balancer was not ready at the time?)

If you install this helm chart, a lot of k8s resources are registered with the k8s api-server: the autohttp's pod, and the proxy-public k8s Service, and more. As part of those resources being registered with the k8s api-server, various controllers start to act in order to make sure that infrastructure represents the resources correctly. A container is created for the autohttp's pod by the Deployment controller after having downloaded the image, and network routing is arranged by some networking controller reading the k8s Service resource.

The crux is that it isn't possible for the helm chart / the autohttps pod to know if all prerequisite infrastructure has become ready so that an external web request from Let's Encrypt to reach the autohttps pod, which is exactly what needs to be possible for an successful startup and TLS cert acquisition from Let's Encrypt using the ACME protocol and an ACME http01 challenge.

To conclude:
- The Helm chart can't decide when everything is setup and timings depends on cloud providers, network speeds, etc.
- The robust solution is to re-attempt an ACME challenge (involving Let's Encrypting sending web request that needs to reach the autohttps pod).
- Re-attempting failed ACME challenges is currently done manually by restarting the autohttps pod.
- Traefik has not previously supported re-attempting an ACME challenge on initial failure, and we have not designed a workaround to detect such failures and forcefully restart traefik yet.

**Comment by pvanliefland** [comment]
I will give it a try!

**Comment by joraff-alt** [comment]
Like others, our attempts to deploy this chart is resulting in Traefik starting a TLS challenge before the service is routing traffic to it.

I was hopeful the Traefik API might have something in an endpoint for the router's TLS status; sadly it doesn't. But I think we've got something to start with in a direction for a functioning startupProbe:

```
cat /etc/acme/acme.json | /tmp/json.sh | egrep '\["default","Certificates",[0-9]+\]\s+\{"domain":\{"main":"escaped_fqdn"\},"certificate":".+$'
```

This depends on [JSON.sh](https://github.com/dominictarr/JSON.sh), which can be installed by wget first run of the probe, and checks that `acme.json` has a certificate value for a certain hostname.

**Comment by consideRatio** [comment]
> Like others, our attempts to deploy this chart is resulting in Traefik starting a TLS challenge before the service is routing traffic to it.

@joraff hmmm... the ACME challenge initiated by Traefik's use of the Go library LEGO from within the autohttps pod, requires the pod to be in a Ready state to complete the http01 challenge (or tls01 challenge i presume, i've not used it). I don't see a simple solution besides making sure that no readiness probe is configured, which would make it not-ready and the challenge fail.

A startupProbe is only an edge case to help the livenessProbe be responsive while being able to tolerate a slower initial startup, it doesn't influence the readiness state of the pod, which in turn influence if traffic is reaching the pod via k8s Services.

If your goal is to await the autohttps pod's true readiness, where it isn't just a Ready pod, but where it has acquired a TLS certificate and can accept traffic, then you can do what we do in this Helm chart's CI system.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/b6688f5482d0f1282afe1586ea52d85f89bae935/ci/common#L36-L41

**Comment by joraff** [comment]
You're right, startupProbe isn't right. A liveness probe will have to be used for this approach.

And of course, we are doing something similar to this in our our CD pipeline (restarting the autohttps rollout), but our intention here, which I should have mentioned, is to distribute a portable terraform module within our organization to deploy this into their own cloud accounts. Not everyone will have or use pipelines so we'd like to not have to rely on external tools.

An even simpler way might be to add a half second initContainer for everyone, in hindsight.

**Comment by consideRatio** [comment]
Hmmm, a livenessProbe is not doing anything to help you either. It is just about restarting pods that fail to respond after a while. You may think a readinessProbe is useful - i did, but that is only reasonable to use if you already have a cert acquired - else, it will make the ACME client unable to receive the traffic it needs to acquire the cert in the first place.

---

Hmmm, I'm not confident about the race condition you are trying to avoid, but this is my perception.

1. When the normal containers of a pod is started, the pod is marked as ready if it doesn't have a readinessProbe configured - this pod don't.
2. When the traefik container starts, its sending out the ACME request, and the clocks starts ticking - all networking, k8s cluster related and cloud provider specific needs to be functional to accept traffic all the way to the pod.

So what needs to happen after the traefik container starts?
1. The implementation of how a k8s Service works with regards to internal routing of traffic that has destined to its associated IP, need to have reacted to the newly started pod has become ready so any traffic sent to the Service IP gets routed to a ready pod (that single pod). A common implementation is that IP tables are configured on the k8s cluster's nodes I think.
2. The implementation of how a k8s Service works with regards to routing traffic from outside the k8s cluster into a k8s Service of a type LoadBalancer needs to be setup in time.

It is not clear to me what kind of race condition you have, but the only way I see you tackling the race condition related to the pod starting up and the routing tables being updated would be to delay the ACME request. To do that, I think you must either configure traefik to do so somehow or perhaps wrap its startup with a slight delay - like "sleep 1 -> start as normal".

If you want to address the latter kind of race condition, for when the autohttps pod starts and when misc ingress routing is setup for the k8s Service of type LoadBalancer, then you can use an initContainer for the autohttps pod as well.

**Comment by joraff** [comment]
In our case it is indeed requirement 2: the azure load balancer takes longer to start forwarding traffic than the tolerance almost every time. Now, granted, we may have a unicorn network setup with lots of peering and gateways which makes networking longer than usual to provision. And most of the other related issues were resolved by removal of the readinessProbe, fortunately. But even so, this could still prove useful - with sane backoffs and delays - to help cover other edge cases. 👍

**Comment by consideRatio** [comment]
> For what its worth, we're consistently getting this on GKE too now. It started inconsistent about 6 mos ago, and now its more frequently a problem than not whenever autohttps needs to be recreated

I could also reproduce this in GKE. I have a workaround that doesn't risk us causing issues by becoming rate limited by Let's Encrypt which is my main concern on how to workaround this: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/2601#issuecomment-1050406952

The best solution is what is also discussed in the issue above I think, an upstream contribution to traefik to be able to configure a delay attempting ACME challenges following traefik startup. That way, we can startup traefik/autohttps pod quickly whenever we do have a cert already but add a delay to attempt cert acquisition as that can often be causing problems.

**Comment by consideRatio** [comment]
I closed #2431 with motivation in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2431#issuecomment-1118505899. In gist, I want us to support solving both delayed container networking setup and external delays such as a LoadBalancer service being slow to setup initially. I also want this support to not risk to being blocked by Let's Encrypt without knowingly restarting the attempt to acquire a cert repeatedly.

**Comment by consideRatio** [comment]
I'm closing this as a `wontfix` as I perceive it as a quite high risk to support making re-attempts to acquire TLS certificates along with introducing complexity in this Helm chart.

The suggested workaround is the following, to configure an initContainer to sleep, so that the attempt to acquire a cert is likeley to succeed no matter if there is trouble with the CNI to setup networking fast enough for the incoming ACME challenge as requested when traefik starts in the main container of the autohttps pod.

```yaml
proxy:
  traefik:
    extraInitContainers:
      # This startup delay can help the k8s container network interface setup
      # network routing of traffic to the pod, which is essential for the ACME
      # challenge submitted by Traefik on startup to acquire a TLS certificate.
      #
      # Sleeping for 7 seconds has been consistently sufficient to avoid issues
      # in GKE when this workaround was explored initially for GKE.
      #
      - name: startup-delay
        image: busybox:stable
        command: ["sh", "-c", "sleep 10"]
```
